core/primitives-core/src/serialize.rs

```diff
@@ -158,109 +158,136 @@ fn test_base_bytes_format() {
     assert_de_error::<Test>("{\"field\":null}");
 }
 
-pub mod u64_dec_format {
+/// Serialises number as a string; deserialises either as a string or number.
+///
+/// This format works for `u64`, `u128`, `Option<u64>` and `Option<u128>` types.
+/// When serialising, numbers are serialised as decimal strings.  When
+/// deserialising, strings are parsed as decimal numbers while numbers are
+/// interpreted as is.
+pub mod dec_format {
     use serde::de;
-    use serde::{Deserialize, Deserializer, Serializer};
+    use serde::{Deserializer, Serializer};
 
-    pub fn serialize<S>(num: &u64, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        serializer.serialize_str(&format!("{}", num))
+    /// Abstraction between integers that we serialise.
+    pub trait DecType: Sized {
+        /// Formats number as a decimal string; passes `None` as is.
+        fn serialize(&self) -> Option<String>;
+
+        /// Constructs Self from a `null` value.  Returns error if this type
+        /// does not accept `null` values.
+        fn try_from_unit() -> Result<Self, ()> {
```

Comment:
nit: `()` is usually what I’d think `unit` is, so this name is potentially slightly confusing I think. Any reason you didn’t call this:

```suggestion
        fn try_from_null() -> Result<Self, ()> {
```

Reply:
serde has `visin_none` and `visit_unit`.  I initially called it `try_from_none` but serde_json calls `visit_unit` when encountering `null` so figured that’s confusing.  I don’t really have strong opinions here.

Reply:
Okay, lets leave as-is then.